Disable default cover images

ThoWagen opened the pull request:
When no cover images are available we don't want to display any image. A solution could be to configure a completely white "no cove" image. However that invisible image can still be interacted with and that can be confusing.

This PR allows to disable default cover images completely. The endpoint returns a 404 status code instead. noCoverAvailableImage and noQRCodeAvailableImage in the config.ini are still providing default images by default.

Besides that jQuery was removed from covers.js.

ThoWagen committed:
disable default cover images

ThoWagen committed:
updated comments

demiankatz commented on config/vufind/config.ini:
```diff
@@ -1318,7 +1318,7 @@ url = "https://api.vlb.de/api/v1/cover/"
 ; This setting controls the image to display when no qrcode is available.
 ; The path is relative to the base of your theme directory.
-;noQRCodeAvailableImage = images/noQRCode.gif
+noQRCodeAvailableImage = images/noQRCode.gif
```
I assume this change was necessary to avoid changing QR code behavior (unless it was unintentional); should the comment be adjusted to address changed behavior?

ThoWagen replied:
Exactly, since "images/noQRCode.gif" was also the default in `QRCode/Loader.php`, this should not change anything.

demiankatz replied:
But the behavior of commenting out the setting has changed, hasn't it? If you do not provide the setting, then NOTHING will be displayed when a QR code is unavailable, correct? If that's the case, maybe we should just add another line of comment, like:

```
; If commented out, no image will be displayed for unavailable QR codes
```

(Again, sorry if I'm misunderstanding the logic here!)

ThoWagen replied:
Sorry for being unclear. But you are right. When the setting is commented out VuFind will return an 404 error for unavailable images. Those cases are catched by JS and the `<img>` elements are hidden.

By "not change anything" I meant that fresh installations will still have "images/noQRCode.gif" as default "no QR code" image. We should probably add some logic that sets the setting in `Upgrade.php`, right?

PS: I added the comment as you suggested.

demiankatz replied:
Yes, adding code to adjust this in VuFind\Config\Upgrader would definitely be a good idea!

ThoWagen replied:
Unfortunately that does not work as I expected. If a setting is commented out in the local configs it will always be overwritten when upgrading. That means having a setting not commented out by VuFind's default and commenting it out for deactivation local does not really work.

We could either leave "noQRCodeAvailableImage" commented out in the config by default but by that change the default behavior. Or we could set `noQRCodeAvailableImage = false` for deactivating the setting instead of commenting it out.

What do you think?

demiankatz replied:
I suspect that this feature is rarely used, if ever, so it's probably not hugely impactful. I think a note in the changelog about the adjustment to default behavior is probably sufficient if there's no straightforward way to make the transition entirely seamless. I'll put a TODO comment on this to write that up when merging, and I'll do some more review and testing of this as soon as time permits.

ThoWagen replied:
We have the same problem for `noCoverAvailableImage`. So we also would have to deactivate that by default or use `noCoverAvailableImage = false` for deactivation instead of commenting it out.

demiankatz replied:
Ahh, I see. I think I like the idea of supporting a false value, since that's a less backward-breaking change (if I'm thinking about this correctly).

ThoWagen replied:
`false` already works for deactivating the setting. So, I just added a test for that and updated the comments in the config.

demiankatz replied:
Great, thank you!

demiankatz commented on module/VuFind/src/VuFind/Exception/CoverUnavailable.php:
```diff
@@ -0,0 +1,43 @@
+<?php
+
+/**
+ * Cover Unavailable Exception
```
Would it be better to call this ImageUnavailable, since it's being used not just for covers but also for QR codes?

ThoWagen replied:
Yes, that makes sense. Done

demiankatz commented on themes/bootstrap3/templates/layout/layout.phtml:
```diff
@@ -130,10 +130,6 @@
 
       $this->headScript()->appendScript(implode("\n", $appendScripts));
-
-      if ($this->config()->ajaxCoversEnabled()) {
```
Does this change render the ajax covers setting meaningless? Maybe I'm missing something, but it seems to me that if we always load covers.js, we always get the ajax functionality. Maybe that is desired, but if so there may be some other bits and pieces to clean up.

ThoWagen replied:
No, this only decides if the covers.js is loaded or not. Now, we also want to include the file event if we don't to use ajax for the covers, because the unavailable images get hidden with js.

If the covers should be loaded via ajax is decided by adding the "ajaxcover" class, e.g. like here https://github.com/vufind-org/vufind/blob/1390c2efd452bf2f934bd6a15d38867a515798a1/themes/bootstrap5/templates/RecordDriver/DefaultRecord/cover.phtml#L19

demiankatz replied:
Thanks for the clarification!

demiankatz reviewed (changes requested):
Thanks, @ThoWagen. I confess that I haven't closely looked at the Javascript changes yet, but first a few questions about everything else...

ThoWagen committed:
renamed exception

ThoWagen committed:
fixed doc comment

ThoWagen committed:
added comment in configs

ThoWagen committed:
updated comment and added testcase for deactivated setting

demiankatz committed:
Merge branch 'dev' into pull-request/disable-default-cover-images

demiankatz reviewed (changes requested):
I think we might want to change the QR code behavior so that it always uses "images/noQRCode.gif" if nothing else is configured; supporting the `noQRCodeAvailableImage = false` option doesn't seem to work well. There's no error handler on the QR code images, so the 404 causes an actual error to display in the browser... and since the QR codes are displayed inside a drop-down, it's pretty weird to open the drop down only to see a broken link inside it. Admittedly, the "no QR code available" image isn't a lot better, but at least it looks intentional.

I'm not proposing that we need to add a custom Javascript error handler for QR codes. This is a rarely-used feature, and a rarely-encountered error condition on top of that. But I think maybe if you change the default in line 71 of the QRCode Loader from `?? null` to `?? 'images/noQRcode.gif'` and remove the comment about setting noQRCodeAvailableImage to false in config.ini, that may be sufficient.

I've also done a code read of your other translations from jQuery to standard JS in the cover code, and it all looks good to me. However, I have not been able to test AJAX cover mode yet since I don't have access to the services that leverage that functionality. I wonder if it's worth building a Demo cover loader to exercise all the possibilities. I'm willing to do that if you think it would be helpful!

ThoWagen committed:
added demo cover loader

ThoWagen committed:
readded default no qr code availabe image

ThoWagen commented:
Having the loading error handler for the record covers actually does not avoid having 404 error messages in the browser's console. Unfortunately, I did not find a way how to fix that. One could check the availability with a separate ajax request in JS before displaying the images. But that would result in double the requests and some delay when loading the images.

But I agree that we don't need to be able to deactivate the unavailable QR code images and that having an extra handler for that would not be worth it, since errors when loading QR codes should be very rare. I implemented you suggestion.

I also added a demo cover loader that selects one of three demo covers from the themes or no image. The selection is based on the checksum of the record ids. So it selects arbitrary results but in a reproducible way.

demiankatz committed:
Add DemoAjax.

demiankatz committed:
Merge branch 'add-demo-covers' into pull-request/disable-default-cover-images

demiankatz committed:
Add DemoAjax version.

demiankatz committed:
Revise comment.

demiankatz commented:
Thanks, @ThoWagen. I thought it would be useful to be able to test the `ajaxcovers = true` setting by adding backlink support to the demo. However, I couldn't figure out a way to make a single Demo driver handle both proxied and AJAX-driven cover cases, so I ended up just writing a separate `DemoAjax` class. Do you think it's worth keeping both of these, or might it be better to just remove the current Demo and replace it with DemoAjax, so the demo we provide covers the greatest number of use cases?

demiankatz commented:
In any case, the good news is that my new demo class proves that your refactored cover logic handles backlinks correctly.

As for avoiding the 404 errors in the console, the best I can think of would be to do something other than a 404 response -- e.g. would it be possible to return a single pixel transparent image with something in the HTTP response that a listener could detect and use to eliminate the image? That might be a little cleaner in some other (unlikely but possible) scenarios like users with Javascript disabled. But I'm not sure if there's really a sensible mechanism to allow this.

ThoWagen commented:
@demiankatz Thanks, I did not think about the backlink functionality. Unfortunately, the DemoAjax cover loader does not work on my system. The reason is, that I run my instance in a container and VuFind can not call it self using via "localhost:someport"  which is provided as "serverurl". I could change my setup to fix that, but others might have the same problem then.

What dou you think about combining Demo and DemoAjax and using `return 'file://' . $path;` in `getUrl` and `return $this->baseUrl . 'themes/' . $cover['theme'] . '/' . $path;` in `getMetadata`? The backlink functionality could also be configured using the `$key` parameter.

> As for avoiding the 404 errors in the console, the best I can think of would be to do something other than a 404 response -- e.g. would it be possible to return a single pixel transparent image with something in the HTTP response that a listener could detect and use to eliminate the image? That might be a little cleaner in some other (unlikely but possible) scenarios like users with Javascript disabled. But I'm not sure if there's really a sensible mechanism to allow this.

Good idea! We can not directly check the e.g. headers of the response in the "load" event. But we can check the size of the loaded image as long as we don't set it via css to some fixed value in the future or in local themes. We can detect the single pixel images and then hide them. This is a bit hacky but we avoid the errors in the console and also probably the servers logs. What do you think about that?

demiankatz commented:
> What dou you think about combining Demo and DemoAjax and using `return 'file://' . $path;` in `getUrl` and `return $this->baseUrl . 'themes/' . $cover['theme'] . '/' . $path;` in `getMetadata`? The backlink functionality could also be configured using the `$key` parameter.

Yes, that sounds reasonable to me. Do you want to do that, or should I? I don't think it's strictly necessary to use `$key` to configure the backlink functionality, since it will essentially work if `ajaxcovers = true` and get ignored otherwise. Might be an interesting avenue for future flexibility, though.

> Good idea! We can not directly check the e.g. headers of the response in the "load" event. But we can check the size of the loaded image as long as we don't set it via css to some fixed value in the future or in local themes. We can detect the single pixel images and then hide them. This is a bit hacky but we avoid the errors in the console and also probably the servers logs. What do you think about that?

That sounds good to me! I thought we already had a 1x1 transparent GIF somewhere in the project, but I can't seem to find it, so we might need to reintroduce that part.

ThoWagen commented:
Great! I will implement the changes tomorrow.

ThoWagen committed:
combined demo and demoajax and hiding one pixel image as default instead of errors

ThoWagen commented:
Done.

I implemented the configurability of the backlink functionality because it it possible to use ajaxcovers without it, even if that is not really useful then.

ThoWagen committed:
added hidden-image.gif

demiankatz commented on module/VuFind/src/VuFind/Content/Covers/Demo.php:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * Selects demo covers or no cover based on the $ids array and returns location information.
+     *
+     * @param array $ids Associative array of identifiers (keys may include 'isbn'
+     *                   pointing to an ISBN object and 'issn' pointing to a
+     *                   string)
```
Looks like IDE auto-formatting got carried away here. Just fixing it via the suggestion mechanism. :-)

```suggestion
     * @param array $ids Associative array of identifiers (keys may include 'isbn'
     * pointing to an ISBN object and 'issn' pointing to a string)
```

demiankatz committed:
Fix comment formatting.

demiankatz committed:
Fix return type.

demiankatz committed:
Merge remote-tracking branch 'origin/dev' into pull-request/disable-default-cover-images

demiankatz committed:
Favor record ID for more consistent generation.

demiankatz committed:
Add Mink test.

demiankatz reviewed (changes requested):
Thanks, @ThoWagen, this is looking great!

I made one small adjustment to the Demo cover provider; I noticed that it generated different covers in AJAX vs. non-AJAX mode, likely due to the two methods passing ID parameters slightly differently and resulting in different crc32 results. I changed it so that if a recordid is provided, that is used in place of the serialized ID array, so that when writing tests, we can better predict behavior (especially if new identifiers are added in the future).

I also went ahead and wrote a Mink test to exercise some of the behavior now that the Demo provider allows it. However, I'm having trouble with the test intermittently failing. On some occasions, when it expects the cover to be hidden, it is not yet hidden. I have tried adding a delay-and-retry loop, but the problem seems not to go away once it has occurred. I'm not sure what is causing this -- a bug in the code, a flaw in my test, or some kind of bad browser behavior. (I think a flaw in my test is the most likely explanation, but I'm not sure what it would be).

Do you mind running this test locally and letting me know if you see the same issue? If so, I'd be interested in whether you have any ideas on how to fix it.

demiankatz commented on themes/bootstrap5/js/covers.js:
```diff
@@ -1,75 +1,111 @@
+        img.remove();
+        source.remove();
+        if (typeof response.data.html !== 'undefined') {
+          container.innerHTML = VuFind.updateCspNonce(response.data.html);
```
One other thing to consider: have you looked at #4065, which eliminates use of innerHTML? Maybe this can be adjusted to match the strategies found there.

ThoWagen replied:
Makes sense. Done

ThoWagen commented:
@demiankatz I also have problems with the test. First of all, when I run chrome not in headless mode the browser's image cache is used and setting the fallback image does not work anymore.

I also have to add some `sleep()` in the test for some cases to work.

The last test case always fails for me even with `sleep` but not when I run it myself by hand. My first guess is that the images is loaded before the "load" event listener is added. 

I will look into it again next week.

demiankatz commented:
Thanks, @ThoWagen! If your theory about the load event handler is correct, then maybe that's actually a potential timing problem that needs to be addressed.

Please let me know if you need me to look deeper into anything on this end. It would be nice if Mink offered a way to clear the browser cache, but I suspect it does not. Maybe there's some way to inject a hash onto image URLs to force things to behave properly, though that may be too much of a hack...

ThoWagen committed:
disable cover image cache in browsers for tests and wait for completed image loading

ThoWagen committed:
Merge branch 'dev' into pull-request/disable-default-cover-images

ThoWagen committed:
fixed cs and tests

ThoWagen commented:
@demiankatz I fixed the tests by adding an configuration that disables browser caches for cover images, by changing the response header and also adding a hash to the url.

I also added an extra 'wait' statement to the test that assures that the images are loaded before the assertions.

That also removed the problem of the hiding not working at all for me.

ThoWagen committed:
using VuFind.setInnerHtml

demiankatz commented on module/VuFind/src/VuFind/Cover/Router.php:
```diff
@@ -48,30 +49,17 @@ class Router implements \Laminas\Log\LoggerAwareInterface
+    public function __construct(
+        protected string $dynamicUrl,
+        protected CoverLoader $coverLoader,
+        protected Config $config
```
laminas-config has just been abandoned; I'm beginning work to remove it from the project. I'd suggest using a plain array here, and using `->toArray` in the factory, so we don't introduce a new dependency on the abandoned project.

demiankatz replied:
It might also make sense to default this to an empty array so that you don't have to modify tests when there is no configuration being tested.

ThoWagen replied:
Done.

Are there any problems with replacing laminas-config with arrays that would be good to keep in mind?

demiankatz replied:
My plan is to create a `VuFind\Config\Config` that provides backward-compatibility with `Laminas\Config\Config` as a first phase. We might decide to keep that in the long term, but I think it's more likely that we'll quickly deprecate it and try to phase it out in favor of arrays. Really, the Laminas Config object was mainly useful for accessing deep configs in older versions of PHP prior to the introduction of null coalescing. Now that we have `??`, it's pretty easy to access deeply-nested array properties as long as you provide a default to fall back on.

ThoWagen replied:
I see, thanks

demiankatz reviewed (changes requested):
Thanks, @ThoWagen, at a glance this looks like it's in good shape now... however, one suggestion based on an unexpected shift in the landscape that occurred last week!

ThoWagen committed:
using array instead of laminas config

demiankatz commented on module/VuFind/src/VuFind/Cover/RouterFactory.php:
```diff
@@ -77,6 +77,7 @@ public function __invoke(
         }
         $coverLoader = $container->get(\VuFind\Cover\Loader::class);
-        return new $requestedName($base, $coverLoader);
+        $config = $container->get(\VuFind\Config\PluginManager::class)->get('config')->toArray()['Content'];
```
Probably worth adding a default fallback for safety (though this section is unlikely to be missing in practice):

```suggestion
        $config = $container->get(\VuFind\Config\PluginManager::class)->get('config')->toArray()['Content'] ?? [];
```

demiankatz reviewed (changes requested):
One more small suggestion, but I'm just going to go ahead and make this change to spare you the trouble. :-)

demiankatz committed:
Update module/VuFind/src/VuFind/Cover/RouterFactory.php

demiankatz reviewed (approved):
Thanks, @ThoWagen -- tests are passing and this all looks good to me now!
